### PR TITLE
code for model1 and model3

### PR DESCRIPTION
**Pull request by mudit1990**
@sarankrishna920 @doodhwala @rohinikapoor
Changes summary
- dal.py : Filter and remove ‘\r’ from email body
- Model1.py: Contains the code for Model1. Architecture - sr_embedding -> linear_layer -> relu_activation -> linear_layer -> predicted_email_emb -> loss
    - sr_embedding is calculated by concatenation of sender and receiver embeddings
    - Loss is calculated as L2 loss between predicted_email_embedding and email_representation. The email representation is obtained by averaging embeddings from pre-trained word2vec model
    - Current modelling has been done only for the 150 core members. From a run on 5k and 10 emails, it can be seen that roughly 25% of the mails are where senders and receivers belong to this core group
    - Train time. For 5k mails ~13secs. For 10k mails ~ 27secs
    - Sample run output for 10 epochs on 10k emails loss in epoch 0 = [89.005974] , loss in epoch 1 = [57.727646] , loss in epoch 2 = [49.35392],  loss in epoch 3 = [45.410088],  loss in epoch 4 = [43.52198] , loss in epoch 5 = [42.04821] , loss in epoch 6 = [40.922516] , loss in epoch 7 = [39.860706] , loss in epoch 8 = [38.792027] , loss in epoch 9 = [38.225792]
- Model3.py: Contains the code for Model3. Architecture - sr_embedding -> linear_layer -> relu_activation -> linear_layer -> predicted_email_emb -> loss
    - One sr_embedding is calculated for each email as a concatenation of sender_emb and average of all receiver_emb present in the mail
    - Everything else is exactly same as model1. This optimization cuts the training time by half and reduces the bias of emails that have a lot of receivers
    - Train time. 5k emails ~7secs and 10k emails ~16secs
    - Sample run output for 10 epochs on 10k emails loss in epoch 0 = [57.731834] , loss in epoch 1 = [39.074894] , loss in epoch 2 = [31.485912],  loss in epoch 3 = [28.263678],  loss in epoch 4 = [26.365534] , loss in epoch 5 = [24.897657],  loss in epoch 6 = [23.555548] , loss in epoch 8 = [21.856457],  loss in epoch 9 = [21.077713]
- utils.py
    - user_id_lookup: A dictionary that stores a mapping of unique_id to email_id. This unique_id is used to lookup the embeddings in nn.Embeddings layer
    - get_userid(): gets the unique_id based on the email. Returns None if the email is not present in the dictionary
    - populate_userid_mapping(): Populates the user_id_lookup dictionary from file employee_id_mapping.csv. The file as of now stores the mapping for 150 core Enron members as per the data

### CODE REVIEW

File: code/model1.py [hunk 1]
```diff
@@ -1,9 +1,91 @@
...
+
+        for epoch in range(self.epochs):
+            epoch_loss = 0.0
+            for i in range(len(emails)):
+                sender_id = utils.get_userid(emails[i, 0])
+                email_word_reps = w2v.get_sentence(emails[i, 2])
+                # if no word_rep was found for any of the words in the emails, ignore this case
+                if len(email_word_reps) == 0:
+                    continue
+                email_rep = self.get_average_rep(email_word_reps)
```
Comment (mohit-surana) [inline code/model1.py]: Would it be better to just call it inline?
If we reuse it for model3's receiver embeddings, then I think the function is justified.

File: code/model1.py [hunk 2]
```diff
@@ -1,9 +1,91 @@
...
+                    optimizer.zero_grad()
+                    recv_id = utils.get_userid(recv)
+                    # if sender or receiver is not an enron email id, we ignore this data point
+                    if sender_id is None or recv_id is None:
+                        continue
+                    # do the forward pass
+                    pred_email_rep = self.forward(autograd.Variable(torch.LongTensor([sender_id])),
+                                                  autograd.Variable(torch.LongTensor([recv_id])))
+                    # compute the loss
+                    loss = loss_criteria(pred_email_rep, autograd.Variable(torch.from_numpy(email_rep)))
```
Comment (mohit-surana) [inline code/model1.py:74]: Some of these function calls can be pulled out of this loop and put into the outer one for speedups

**Review by mohit-surana — APPROVED** [review]
I think this looks great and can be merged!
Really helpful comments made reviewing the code much easier. I think we all should this from now on.

:shipit:

File: code/model1.py (hunk 1 above)
Comment (mudit1990) [reply code/model1.py]: done
